tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java

```diff
@@ -294,25 +298,82 @@ private JsonNode getCombinedStatus() {
     }
 
     /**
-     * Create a task.
-     *
-     * @param id                    The ID of the task to create.
-     * @param spec                  The specification of the task to create.
+     * Creates and schedules a task.
      */
-    public void createTask(final String id, TaskSpec spec)
+    public void createAndScheduleTask(TaskDetail task)
             throws Throwable {
+        ManagedTask managedTask = createTask(task);
+
+        long delayMs = scheduleTask(managedTask);
+        log.info("Created a new task {} with spec {}, scheduled to start {} ms from now.", task.id, task.spec, delayMs);
+    }
+
+    /**
+     * Atomically creates and schedules the given tasks. If any task fails creation, none get scheduled.
+     */
+    public void createAndScheduleTasksAtomic(List<TaskDetail> tasks) throws BadRequestException, RequestConflictException {
+        List<ManagedTask> managedTasks = new ArrayList<>();
+        if (tasks.stream().map(task -> task.id).collect(Collectors.toSet()).size() < tasks.size())
+            throw new RequestConflictException("Duplicate task IDs given");
+
+        for (TaskDetail task: tasks) {
+            try {
+                managedTasks.add(createTask(task));
+            } catch (Throwable e) {
+                log.info("Failed to create createTask(id={}, spec={}) error:", task.id, task.spec, e);
+            }
+        }
+
+        for (ManagedTask task: managedTasks) {
+            long delayMs = scheduleTask(task);
+            log.debug("Created a new task {} with spec {}, scheduled to start {} ms from now.", task.id, task.spec, delayMs);
+        }
+
+        log.info("Created {} tasks.", managedTasks.size());
+    }
+
+    private long scheduleTask(ManagedTask task) {
+        tasks.put(task.id, task);
+        long delayMs = task.startDelayMs(time.milliseconds());
+        task.startFuture = scheduler.schedule(executor, new RunTask(task), delayMs);
+        return delayMs;
+    }
+
+    private ManagedTask createTask(TaskDetail task) throws Throwable {
         try {
-            executor.submit(new CreateTask(id, spec)).get();
+            if (task.id.isEmpty()) {
+                throw new InvalidRequestException("Invalid empty ID in createTask request.");
+            }
+            ManagedTask duplicateTask = tasks.get(task.id);
+            if (duplicateTask != null) {
+                log.info("Task {} already exists with spec {}", duplicateTask.id, duplicateTask.spec);
+                throw new RequestConflictException("Task ID " + duplicateTask.id + " already exists");
```

Comment:
I commented earlier about this, but re-creating the same task with the same spec should not cause an exception.

Reply:
We are robust either way, the task is created. The question is how we want to inform the caller of what has happened - a `409` or `201` both do the same thing, they tell him that a task with that ID already exists on the server

Reply:
Check TROGDOR.md.

> All Trogdor RPCs are idempotent except the shutdown requests.  Sending an idempotent RPC twice in a row has the same effect as sending the RPC once.

Because the request is idempotent, sending it twice has the same effect, including the same result code.

Reply:
See my other message, I don't think it has to have the same status code. However, I agree that it is unexpected to receive an error if the request succeeded. But a 200/204 are possible options.